design/StableMemory.md

```diff
@@ -35,73 +30,204 @@ ic0.stable_write : (offset : i32, src : i32, size : i32) -> ();             // *
-* Returns ref contents from stable memory as a message reply would require redundant de- and re-serialization unless we optimize this to a bulk read from stable storage to memory.
+```
+module StableMemory {
+  Nat8 : module {
+    write : (offset : Nat8, n : Nat8) -> ();
+    read : (offset : Nat8) -> Nat8;
+  }
+  // etc for all scalar prim types.
+  ...
+  Blob : {
+    write : (offset : Nat32, b : Blob) -> ();
+    read : (offset : Nat32, size : i32) -> Blob
+  }
+}
+```
+(I think the compiler will still optimize these nested calls to known
+function calls, but it would be worth checking).
 
-# Maintaining existing Stable variables.
+# Maintaining existing Stable Variables.
 
 Stable memory is currently hidden behind the abstraction of stable
 variables, which we will still need to maintain. The current
 implementation of stable variables stores all variables as a
 Candid(ish) record of _stable_ fields, starting at stable memory address 0 with
 initial word encoding size (in bytes?) followed by contents.
 
-Ideally, we would extend this so all user-defined StableMemory is stored at a low address, with
-stable _variable_ data stored just beyond the currently used StableMemory
-content on canister_pre_upgrade and canister_post_upgrade. That way the StableMemory area need not move, with stable variables simply serialized and appended in `canister_pre_upgrade` and deserialized and discarded in `canister_post_upgrade`, leaving the manual StableMemory unchanged.
+Starting from a clean slate, we would extend this so all user-defined StableMemory is
+stored at a low address, with _stable variable_ data stored just
+beyond the currently used StableMemory content on canister_pre_upgrade
+and canister_post_upgrade. That way the StableMemory area need not
+move, with stable variables simply serialized and appended in
+`canister_pre_upgrade` and deserialized and discarded in
+`canister_post_upgrade`, leaving the manual StableMemory unchanged.
 
-Assuming stable variables always require non-zero bytes to encode, we should be able to devise a backwards compatible scheme for upgrading from pre-StableMemory canisters to post-StableMemory canisters. Details to be designed.
+For backwards compatibility reasons, we can't do that.
 
+Luckily, stable variables always require non-zero bytes to encode, we
+should be able to devise a backwards compatible scheme for upgrading
+from pre-StableMemory canisters to post-StableMemory
+canisters, as follows.
 
+During execution, abstract stable memory (StableMemory) is aligned
+with IC stable memory, at address 0, for reasonable efficiency (apart
+from maintaining `max`).
 
+During upgrade, we compute the size of the stable variable blob, shift
+that portion of StableMemory to the end of user memory (growing if
+necessary) and write the old format into the space just freed,
+followed by the size of StableMem.
```

Comment:
This may be more complicated and expensive than necessary, I think. Why shift all the memory? As you observe, you can distinguish old layout from new layout by checking whether the byte at address 0 is non-zero. So isn't it sufficient to just back-up this one byte somewhere else in the new scheme and null it out for upgrade?

Along with the idea of keeping grow explicit and dropping `max`, stable memory could then be used like this during upgrade:
```
[0]  0
[1..N-1]  StableMemory bytes
[N..M-7]  serialised stable vars (+ possible page padding)
[M-6..M-3]  value N/64Ki = StableMemory.size()
[M-2]  backed-up StableMemory byte 0
[M-1]  version byte
```
where

M = physical size of stable memory (as seen by system)
N = logical size of stable memory (as seen through StableMemory interface)

both being multiples of 64Ki, i.e., page boundaries.

Reply:
I'm not shifting all memory just the portion occupied by stables vars but I'll consider this scheme instead. However, the first word (not byte) must be 0x0.

Reply:
> However, the first word (not byte) must be 0x0.

I was wondering about that. Isn't the first byte always non-zero in a Candid blob?

Reply:
I think we write the size of the encoding plus the encoding itself at the moment, so the first 4 bytes aren't candid at all. But I'll check.

Reply:
confirmed